### PR TITLE
Filter more attributes

### PR DESCRIPTION
**Pull request by TheCovenant**

### CODE REVIEW

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/FeatureAttribute.ts [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 
 // TypeScript
 
 export interface FeatureAttribute {
   readonly name: string;
-  readonly isNumeric: boolean;
-  readonly isString: boolean;
+  readonly typeOf: TypeOfFeatureAttribute;
   readonly filterType: FilterType;
+  readonly fieldType: FieldType;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/FeatureAttribute.ts:13]: I'd call this dataType and DataType, more conventional name.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/FrequencyFeatureAttribute.ts [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+import {TypeOfFeatureAttribute} from "../states/map/TypeOfFeatureAttribute";
+import {FieldType} from "../states/map/FieldType";
 
 export class FrequencyFeatureAttribute implements FeatureAttribute {
   static readonly instance = new FrequencyFeatureAttribute();
 
-  readonly isNumeric = true;
   readonly name = FeatureAttributeName.transmissionPower;
-  readonly isString = false;
+  readonly typeOf = TypeOfFeatureAttribute.NUMBER;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/FrequencyFeatureAttribute.ts:11]: Confusing why you have this and dataType/fieldType. Just combine them? Either way, typeOf is way too generic a name.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/LabelFeatureAttribute.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import {FeatureAttribute} from "./FeatureAttribute";
+import {FeatureAttributeName} from "../states/map/FeatureAttributeName";
+import {FilterType} from "../states/map/FilterType";
+import {TypeOfFeatureAttribute} from "../states/map/TypeOfFeatureAttribute";
+import {FieldType} from "../states/map/FieldType";
+
+export class LabelFeatureAttribute implements FeatureAttribute {
+  static readonly instance = new LabelFeatureAttribute();
+  readonly isNumeric = false;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/LabelFeatureAttribute.ts:9]: Don't need isNumeric and isString?

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/LocalityFeatureAttribute.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import {FeatureAttribute} from "./FeatureAttribute";
+import {FeatureAttributeName} from "../states/map/FeatureAttributeName";
+import {FilterType} from "../states/map/FilterType";
+import {TypeOfFeatureAttribute} from "../states/map/TypeOfFeatureAttribute";
+import {FieldType} from "../states/map/FieldType";
+
+export class LocalityFeatureAttribute implements FeatureAttribute {
+  static readonly instance = new LocalityFeatureAttribute();
+
+  readonly isNumeric = false;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/LocalityFeatureAttribute.ts:10]: Don't need isNumeric and isString?

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/RegionsFeatureAttribute.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import {FeatureAttribute} from "./FeatureAttribute";
+import {FeatureAttributeName} from "../states/map/FeatureAttributeName";
+import {FilterType} from "../states/map/FilterType";
+import {TypeOfFeatureAttribute} from "../states/map/TypeOfFeatureAttribute";
+import {FieldType} from "../states/map/FieldType";
+
+export class RegionsFeatureAttribute implements FeatureAttribute {
+  static readonly instance = new RegionsFeatureAttribute();
+
+  readonly isNumeric = false;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/RegionsFeatureAttribute.ts:10]: Same?

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterPanel/FilterPanel.tsx [hunk 6]
```diff
@@ -11,7 +11,7 @@ import {useSelector, connect} from "react-redux";
 import {MapState} from "../../states/map/MapState";
 import {RootState} from "../../states/root/RootState";
 import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
-import {FilterSliders} from "../filterSliders/FilterSliders";
+import {FilterSliders} from "../filterSliders/FilterComponent";
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterPanel/FilterPanel.tsx:14]: File name different from component name?!

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterComponent.tsx [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    );
+    dispatch(
+      setFilter(
+        filterIndexOfAttribute,
+        "type",
+        getFeatureAttributeByName(attributeName).filterType
+      )
+    );
+    switch (getFeatureAttributeByName(attributeName).typeOf) {
+      case TypeOfFeatureAttribute.NUMBER: {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterComponent.tsx:87]: I see how you're using this TypeOfFeatureAttribute, sortof.

This is really where the strategy pattern is supposed to shine. Right now you're using FeatureAttribute as a set of flags indicating how to switch on certain behavior. But the strategy pattern is really about polymorphic delegation. Can you turn this switch into a delegation to the strategy? i.e.,

const featureAttribute = getFeatureAttributeByName(attributeName);
featureAttribute.getInitialFilterActions(...)

You could have two levels of strategy singletons, like:

featureAttribute.filterType.getInitialFilterActions(...).

in order to reuse code.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterComponent.tsx [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+  const returnFilterComponent = (
+    filterIndexOfAttribute: number,
+    attributeName: string,
+    stateOfAttribute: MapFeatureAttributeState
+  ) => {
+    switch (getFeatureAttributeByName(attributeName).typeOf) {
+      case TypeOfFeatureAttribute.NUMBER: {
+        stateOfAttribute = stateOfAttribute as MapNumericFeatureAttributeState;
+        return (
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterComponent.tsx:117]: Following from the above, you could do:
featureAttribute.filterType.render(stateOfAttribute, ...) -> React.ReactNode
But I think I prefer the way you have it right now, keeping the actual JSX in one place.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterComponent.tsx [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  //Get the featureTypeState of the appropriate featureType which was passed in by FilterPanel as a prop
+  const featureTypeState = state.featuresByType[featureType].featureTypeState;
+
+  //attributeStatesOfFeatureType is now an object, with each of its properties being the state of
+  //an attribute of that FeatureType.
+  //i.e. attributeStatesOfFeature = {frequency: {min:0, max:100, filterIndex: 0}, tranmsmissionPower: {min:0, max:20, filterIndex: 1}}
+  const attributeStatesOfFeatureType =
+    state.featuresByType[featureType].attributeStates;
+
+  const fake_state: any = useSelector(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/filterSliders/FilterComponent.tsx:197]: rly

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 10]
```diff
@@ -105,11 +102,11 @@ const MapImpl: React.FunctionComponent = () => {
   //if there are no states loaded
   if (state.features.length === 0) {
     //if the data variable has been loaded
-    if (initialFeaturesQueryResult.data) {
+    if (keplerState.map) {
       // Not tracking any features yet, add the boroughs we loaded
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:106]: Love this ancient comment.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 11]
```diff
@@ -105,11 +102,11 @@ const MapImpl: React.FunctionComponent = () => {
...
       dispatch(
         addMapFeatures(
-          initialFeaturesQueryResult.data.features.map(feature => ({
+          stateJSON.map(feature => ({
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:109]: In general I'd prefer to encapsulate this, but it's OK for now.
https://www.apollographql.com/docs/react/caching/cache-interaction/

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 12]
```diff
@@ -208,7 +205,7 @@ const MapImpl: React.FunctionComponent = () => {
             !hasDirtyFeatures(state.featuresByType)
           ) {
             //Dispatch the addFilter action 3 times (1 for each of frequency, timeStamp, transmissionPower)
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:207]: Comment is out of date?

File: app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts [hunk 13]
```diff
@@ -11,34 +14,66 @@ export const updateAttributeStatesOfFeatureType = (
...
-        attributeStateOfFeatureType.max === null
-      ) {
-        //New max of the attribute wil be the feature's value for the attribute
-        attributeStateOfFeatureType.max = addedFeature[attributeKey] as number;
-        //New min of the attribute wil be the feature's value for the attribute
-        attributeStateOfFeatureType.min = addedFeature[attributeKey] as number;
-        continue;
+    switch (getFeatureAttributeByName(attributeName).typeOf) {
+      //If the attribute is numeric
+      case TypeOfFeatureAttribute.NUMBER: {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/reducerFunctions/updateAttributeStatesOfFeatureType.ts:22]: Another good place to delegate to the strategy. featureAttribute...buildInitialState(...)

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState.ts [hunk 14]
```diff
@@ -8,7 +8,7 @@ This is important info that will be used for setting filter values.
 Unsure, if I should call this FeatureTypeAttributeState instead.
 Or maybe 'AttributeOfFeatureTypeState'
 */
-export interface MapFeatureAttributeState {
+export interface MapNumericFeatureAttributeState {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState.ts:11]: I like this specialization. Do all of these fields have to be null? Can you make the object undefined until you can populate everything?

In general, if two fields must be populated together or neither populated, then you should make them a separate interface + single member on the parent interface. I suspect that's the case with min and max here. Either you have them both or you have neither.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/initialMapState.ts [hunk 15]
```diff
@@ -33,12 +34,22 @@ Object.values(FeatureType).map(type => {
...
     const FeatureAttribute = getFeatureAttributeByName(attributeName);
-    if (FeatureAttribute.isNumeric) {
-      featuresByType[type].attributeStates[attributeName] = {
-        min: null,
-        max: null,
-        filterIndex: null,
-      };
+    switch (FeatureAttribute.typeOf) {
+      case TypeOfFeatureAttribute.NUMBER: {
+        featuresByType[featureType].attributeStates[attributeName] = {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/initialMapState.ts:39]: Another delegation..

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/FrequencyFeatureAttribute.ts [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+import {TypeOfFeatureAttribute} from "../../states/map/TypeOfFeatureAttribute";
+import {KeplerFilterType} from "../../states/map/KeplerFilterType";
+import {KeplerFieldType} from "../../states/map/KeplerFieldType";
+import {Dispatch} from "redux";
+import {buildInitialFeatureAttributeStateNumeric} from "../functions/buildInitialFeatureAtttributeState/buildInitialFeatureAttributeStateNumeric";
+import {MapFeature} from "../../states/map/MapFeature";
+import {updateNumericAttributeStateOfFeatureType} from "../functions/updateAttributeStateOfFeatureType/updateNumericAttributeStateOfFeatureType";
+
+export class FrequencyFeatureAttribute implements FeatureAttribute {
+  updateAttributeStatesOfFeatureType(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/FrequencyFeatureAttribute.ts:14]: If this is all the delegation is, have an abstract parent class NumericFeatureAttribute?

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/LocalityFeatureAttribute.ts [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+import {setInitialFiltersString} from "../functions/setInitialFilters/setInitialFiltersString";
+import {MapFeatureAttributeState} from "../../states/map/MapFeatureAttributeState/MapFeatureAttributeState";
+import {Dispatch} from "redux";
+import {buildInitialFeatureAttributeStateString} from "../functions/buildInitialFeatureAtttributeState/buildInitialFeatureAtttributeStateString";
+import {MapFeature} from "../../states/map/MapFeature";
+import {updateStringAttributeStateOfFeatureType} from "../functions/updateAttributeStateOfFeatureType/updateStringAttributeStateOfFeatureType";
+
+export class LocalityFeatureAttribute implements FeatureAttribute {
+  static readonly instance = new LocalityFeatureAttribute();
+  updateAttributeStatesOfFeatureType(
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/LocalityFeatureAttribute.ts:15]: StringFeatureAttribute parent class?

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState.ts [hunk 18]
```diff
@@ -8,8 +8,12 @@ This is important info that will be used for setting filter values.
...
+  range: MapFeatureAttribtueNumericRange;
+  filterIndex: number | null;
+}
+
+interface MapFeatureAttribtueNumericRange {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState.ts:16]: typo

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  ): void {
+    setInitialFiltersNumeric(
+      filterIndexOfAttribute,
+      this.name,
+      stateOfAttribute,
+      this.KeplerFilterType,
+      dispatch
+    );
+  }
+  abstract readonly name: FeatureAttributeName;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts:47]: I don't think you need to redeclare these if they're in FeatureAttribute?

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/TimeStampFeatureAttribute.ts [hunk 20]
```diff
@@ -1,53 +1,11 @@
...
-    setInitialFiltersNumeric(
-      filterIndexOfAttribute,
-      this.name,
-      stateOfAttribute,
-      this.KeplerFilterType,
-      dispatch
-    );
-  }
-  readonly name = FeatureAttributeName.timestamp;
+export class TimestampFeatureAttribute extends NumericFeatureAttribute {
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/TimeStampFeatureAttribute.ts:6]: class name inconsistent with file name. Usual spelling is Timestamp.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/functions/setInitialFilters/setInitialFiltersString.ts [hunk 21]
```diff
@@ -1,8 +1,8 @@
 import {MapFeatureAttributeState} from "../../../states/map/MapFeatureAttributeState/MapFeatureAttributeState";
 import {setFilter} from "kepler.gl/actions";
-import {MapNumericFeatureAttributeState} from "../../../states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState";
 import {KeplerFilterType} from "../../../states/map/KeplerFilterType";
 import {Dispatch} from "redux";
+import {MapStringFeatureAttributeState} from "../../../states/map/MapFeatureAttributeState/MapStringFeatureAttributeState";
 
 export const setInitialFiltersString = (
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/functions/setInitialFilters/setInitialFiltersString.ts:7]: You shouldn't need these freestanding functions, or? This is just the implementation of StringFeatureAttribute.setInitialFilters. If it's not used elsewhere move this code there.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx [hunk 22]
```diff
@@ -271,7 +270,7 @@ const MapImpl: React.FunctionComponent = () => {
           //A query is not in progress AND if the length of the previous query was capped by the limit variable that was passed...
           if (
             featureLoadingState.latestQueryLength === limit &&
-            (featureLoadingState.offset < 2000 || !DEBUG)
+            (featureLoadingState.offset < 5000 || !DEBUG)
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/map/Map.tsx:273]: Magic numbers should go in const's at file level so it's clear what they are. DEBUG_FEATURES_PER_TYPE_MAX = 5000 or something like this.

File: app/geo/gui/src/ts/twxplore/gui/geo/components/mapPage.tsx/MapPage.tsx [hunk 23]
```diff
@@ -0,0 +1,156 @@
+import * as React from "react";
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/components/mapPage.tsx/MapPage.tsx:1]: Directory names shouldn't have extensions.
Usual nomenclature for this type of component is a "layout". MapLayout.

Comment (gordom6) [reply app/geo/gui/src/ts/twxplore/gui/geo/components/mapPage.tsx/MapPage.tsx:1]: You could put MapLayout in components/map/, the component subdirectories are for functionally similar components.

File: app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState.ts [hunk 24]
```diff
@@ -9,11 +9,11 @@ Unsure, if I should call this FeatureTypeAttributeState instead.
...
 */
 export interface MapNumericFeatureAttributeState {
-  range: MapFeatureAttribtueNumericRange;
+  range: MapFeatureAttributeNumericRange;
   filterIndex: number | null;
 }
 
-interface MapFeatureAttribtueNumericRange {
+interface MapFeatureAttributeNumericRange {
   min: number | null;
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/states/map/MapFeatureAttributeState/MapNumericFeatureAttributeState.ts:17]: This is incorrect. The semantics are that both min and max must be set or neither must be set. So what you really want is 

MapFeatureAttributeNumericRange { min: number; max: number; }
then above
range: MapFeatureAttributeNumericRange | null

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts (hunk 19 above)
Comment (TheCovenant) [reply app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts:47]: It seems so. Else the class won't implement FeatureAttribute as declared.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts [hunk 25]
```diff
@@ -17,36 +15,66 @@ export abstract class NumericFeatureAttribute implements FeatureAttribute {
...
+    //Specifying the attribute state of the addedFeature's FeatureType to be updated (e.g. the attribute state of 'frequency' of Transmissions)
+    let attributeStateOfAttributeOffFeatureType = attributeStatesOfFeatureType[
+      attributeName
+    ] as MapNumericFeatureAttributeState;
+    //If this is the first time coming across this attribute for the addedFeature's FeatureType
+    if (!attributeStateOfAttributeOffFeatureType) {
+      //Give the MapNumericAttributeState min/max the addedFeature's value for the attribute
+      attributeStatesOfFeatureType[attributeName] = {
+        range: {
+          min: addedFeature[attributeKey] as number,
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts:29]: I see a bunch of duplicate expressions here, like addedFeature[attributeKey] as number.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts [hunk 26]
```diff
@@ -17,36 +15,66 @@ export abstract class NumericFeatureAttribute implements FeatureAttribute {
...
+        filterIndex: null,
+      };
+
+      return;
+    }
+    //If this is NOT the first time coming across this attribute for the addedFeature's FeatureType
+    //Compare attribute value of addedFeature to the min found in the attribute state. Set new min if necessary.
+    if (
+      (addedFeature[attributeKey] as number) <
+      attributeStateOfAttributeOffFeatureType.range!.min
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/NumericFeatureAttribute.ts:41]: Another duplicate expression.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/StringFeatureAttribute.ts [hunk 27]
```diff
@@ -17,36 +15,53 @@ export abstract class StringFeatureAttribute implements FeatureAttribute {
...
-  }): void {
-    buildInitialFeatureAttributeStateString(
-      attributeStatesOfFeatureType,
-      this.name
-    );
+    const attributeName = this.name;
+    const attributeKey = this.name as keyof MapFeature;
+    //Specifying the attribute state of the addedFeature's FeatureType to be updated (e.g. the state of 'frequency' of Transmissions)
+
+    let attributeStateOfAttributeOfFeatureType = attributeStatesOfFeatureType[
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/StringFeatureAttribute.ts:22]: const? ESLint should have caught it.

File: app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/StringFeatureAttribute.ts [hunk 28]
```diff
@@ -17,36 +15,53 @@ export abstract class StringFeatureAttribute implements FeatureAttribute {
...
+      };
+      attributeStateOfAttributeOfFeatureType = attributeStatesOfFeatureType[
+        attributeName
+      ] as MapStringFeatureAttributeState;
+    }
+    //add the value of addedFeature's attribute into the values list of the attribute state if the value is not null and is not already included.
+    if (
+      addedFeature[attributeKey] !== null &&
+      !attributeStateOfAttributeOfFeatureType.values.includes(
+        addedFeature[attributeKey] as string
```
Comment (gordom6) [inline app/geo/gui/src/ts/twxplore/gui/geo/attributeStrategies/classes/StringFeatureAttribute.ts:40]: More duplicate expressions

**Pull request merged**